ha/lock.go

```diff
@@ -0,0 +1,127 @@
+package ha
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/uuid"
+)
+
+const (
+	renewalFactor = 0.7
+	waitFactor    = 1.1
+)
+
+type lock interface {
+	Acquire(ctx context.Context, callerID string) (string, error)
+	Release(ctx context.Context) error
+	Renew(ctx context.Context) error
+}
+
+type HAController struct {
+	ID            string
+	renewalPeriod time.Duration
+	waitPeriod    time.Duration
+	randomDelay   time.Duration
+
+	logger log.Logger
+	lock   lock
+}
+
+func NewHAController(l lock, logger log.Logger, lease time.Duration) *HAController {
+	logger = logger.Named("ha_mode")
+
+	rn := rand.New(rand.NewSource(time.Now().Unix())).Intn(100)
+	hac := HAController{
+		lock:          l,
+		logger:        logger,
+		renewalPeriod: time.Duration(float64(lease) * renewalFactor),
+		waitPeriod:    time.Duration(float64(lease) * waitFactor),
+		ID:            uuid.Generate(),
+		randomDelay:   time.Duration(rn) * time.Millisecond,
+	}
+
+	return &hac
+}
+
+func (hc *HAController) Start(ctx context.Context, protectedFunc func(ctx context.Context)) error {
+	hc.logger.Named(hc.ID)
+
+	// To avoid collisions if all the instances start at the same time, wait
+	// a random time before making the first call.
+	hc.wait(ctx)
+
+	waitTicker := time.NewTicker(hc.waitPeriod)
+	defer waitTicker.Stop()
+
+	for {
+		hc.logger.Debug("attempting to acquire lock")
+		lockID, err := hc.lock.Acquire(ctx, hc.ID)
+		if err != nil {
+			// TODO: What to do with fatal errors?
+			hc.logger.Error("unable to get lock", err)
+		}
+
+		if lockID != "" {
+			hc.logger.Debug("lock acquired, ID", lockID)
+			funcCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// Start running the lock protected function
+			go protectedFunc(funcCtx)
```

Comment:
I've been thinking about what would happen if this goroutine were to exit prematurely. I think we would end-up with an instance holding the lock forever but not actually doing anything? 🤔 

I'm not sure how to solve this, maybe some kind of health check mechanism so we only renew the lease if the function is still active? But maybe that's too much overhead for now, and even health checks can lie.

Reply:
we can think about inverting the order and calling the lock as a routine and the main service as we always did?

Reply:
Also, the autoscaler can cancel its context on error and we can catch that on the ha controller to return, what do you think? @lgfa29